### PR TITLE
DEV: add environment.yml file for development with conda/mamba

### PR DESCRIPTION
**Pull request by rgommers**
This allows installing all development dependencies with a single command in a new environment.

### CODE REVIEW

File: environment.yml
Comment (tupui) [inline environment.yml:7]: NumPy is missing no? Or I am missing something?

**Review by tupui — APPROVED** [review]
Nice! Better than the listing we had.

File: environment.yml
Comment (rgommers) [reply environment.yml:7]: It got pulled in by Matplotlib, but yeah .... oops. Fixed now.

**Comment by trallard** [comment]
@rgommers 🤔  shall we then use this `environment.yml` for the Docker image creation (once this gets merged of course)?
This will reduce the need to keep this and the one in `tools/docker_dev` up to date

**Comment by rgommers** [comment]
Ah good point, I already forgot we had another `environment.yml` there. Yes, I think it's better to have it in the root of the repo, easier to discover and will therefore also be kept better up to date. Let me unify the two, there's some differences in packages.

**Comment by tupui** [comment]
I though both served different purposes and so did not have the same installed packages. So we might still need to have both.

**Comment by rgommers** [comment]
> I though both served different purposes and so did not have the same installed packages. So we might still need to have both.

Maybe slightly different - but in the end both are used to set up a SciPy development environment.

**Comment by rgommers** [comment]
After reviewing `tools/docker_dev/environment.yml`, I added `conda-build` and `flake8`. The `ipython` pinning is no longer needed, because latest IPython has fixed its tab completion.

I think it's best to leave possible removal of `tools/docker_dev/environment.yml` to your Gitpod PR @trallard. I think the main difference is that an `environment.yml` file for general use should not use pinned packages, because those go out of date very easily. We only use the necessary pinnings in CI configs also, and not everything. What would be useful is to use this `environment.yml` in a regular CI job, to detect possible test failures (more likely than build failures) - but that's for another PR.

The complete solve takes about 1 minute with conda, and less with mamba, so the added packages are fine.

File: environment.yml
Comment (andyfaff) [inline environment.yml:15]: Should we add pythran dependencies, or leave it to the resolver to figure that out?

Comment (rgommers) [reply environment.yml:15]: Indirect dependencies are always for the resolver to deal with.

**Comment by rgommers** [comment]
No more comments, everyone seemed in favor - merging.

**Pull request merged**
